server/src/main/java/org/elasticsearch/search/SearchHit.java

```diff
@@ -426,16 +476,30 @@ public DocumentField field(String fieldName) {
      * were required to be loaded.
      */
     public Map<String, DocumentField> getFields() {
-        return fields == null ? emptyMap() : fields;
+        Map<String, DocumentField> fields = new HashMap<>();
+        fields.putAll(metaFields);
+        fields.putAll(documentFields);
+        return fields;
     }
 
     // returns the fields without handling null cases
     public Map<String, DocumentField> fieldsOrNull() {
-        return fields;
+        return getFields();
     }
 
     public void fields(Map<String, DocumentField> fields) {
-        this.fields = fields;
+        this.metaFields.clear();
+        this.documentFields.clear();
+        if (fields == null) {            
```

Comment:
While `null` is allowed right now, I don't think we ever actually pass it. I see 3 non-test uses of this method, all of which pass a non-null map. I think you can add a requiresNonNull here.